README.md
# node-rand-token

Generate random tokens from your choice of randomness.

[![NPM](https://nodei.co/npm/rand-token.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/rand-token/)

[![Build Status](https://travis-ci.org/sehrope/node-rand-token.svg?branch=master)](https://travis-ci.org/sehrope/node-rand-token)


# Installation

Add it to your node.js project via:

    npm install rand-token --save

# Usage

    // Create a token generator with the default settings:
    var randtoken = require('rand-token');

    // Generate a 16 character alpha-numeric token:
    var token = randtoken.generate(16);

    // Use it as a replacement for uid:
    var uid = require('rand-token').uid;
    var token = uid(16);

    // Generate mostly sequential tokens:
    var suid = require('rand-token').suid;
    var token = suid(16);

# TypeScript usage

This module includes type definition:

```typescript
import { uid, suid } from 'rand-token';

const token = uid(12);
const otherToken = suid(16);
```

# Defaults

The default set of allowed characters is all alpha-numeric characters. Specifically, lower case a through z, upper case A through Z, and the number 0 through 9. This gives you `(26 + 26 + 10)` = `62` possibilities per character.

Using 8 character random tokens will give you a possible token space of 62^8 = 2.18 x 10^14 ~= 2^47

Using 12 character random tokens will give you a possible token space of 62^12 = 3.22 x 10^21 ~= 2^71

Using 16 character random tokens will give you a possible token space of 62^16 = 4.76 x 10^28 ~= 2^95

# Functions

## uid(size)
Uses the default generator to generate a token of `size` characters.

## suid(size, [epoch], [prefixLength])
Uses the default generator to generate *mostly* sequential ids that can be compared with the usual string less-than/greater-than operators. By mostly, it means that a second execution of this function within the same millisecond may generate an id that is less than the second. The same holds true for ids generated from separate node processes. There's no coordination whatsover. This is meant to be used for situations where being able to sort tokens would be convenient, but not strictly required.

The generated tokens are of `size` characters prefixed with the time since the given epoch in base62 padded to `prefixLength` characters. The `epoch` parameter should be the unix time offset in milliseconds. The default `epoch` is 2000-01-01T00:00:00+00:00 and the default `prefixLength` is 8 characters.

__NOTE:__ The prefix length is in addition to the `size` parameter. Calling this function as `suid(16)` will return a 24-character token back (8 + 16).

## generate(size, [chars])

Generates a token of `size` characters using either the specified `chars` or the default for the generator. 

If `chars` is specified it will be treated as an array of characters to use. Each character in the list has an equal chance of being used so if a character is repeated twice, it will appear twice as often in the randomly generated tokens.

__Note__: Unlike the `generator(...)` function this function does not treat any string values of `chars` to be special and all values will be simply treated as an array of possible characters.

## generator([options])

Creates a custom token generator.

Available options:

* `source` - source of random bytes
    This should be either a string or a synchronous function with the signature `(size: number): Buffer`.

    The following string values are also accepted:

    * `default` - This is a synonym for using the default of `crypto.pseudoRandomBytes`. You do not need to specify `default` and can simply create the generator without this option specified to get the same effect.

    * `crypto` - This is a synonym for [`crypto.randomBytes`](http://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback). Note that this may throw an error during use if there is not enough entropy.

    * `math` - This is a synonym for using the Math.random() function.

* `chars` - string representing the list of allowed characters to use for `generate(size)` calls.

    The following string values have special meaning:

    * `default` - default set of token characters which is numbers, lower case letters, and upper case letters (i.e. `0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`)
    * `a-z` || `alpha` - lower case characters (i.e. `abcdefghijklmnopqrstuvwxyz`)
    * `A-Z` || `ALPHA` - upper case characters (i.e. `ABCDEFGHIJKLMNOPQRSTUVWXYZ`)
    * `0-9` || `numeric` - numbers only (i.e. `0123456789`)
    * `base32` - use characters from the base32 alphabet, specifically `A-Z` and `2-7`

    Any other string value will be treated as an array of characters to use. 

    Each character in the list has an equal chance of being used. For example if a character is repeated twice, it will appear twice as often in randomly generated tokens. `chars` may be at most 255 characters long.

# Examples

To replace usage of uid with the default generator (alpha-numeric):

    // Create a token generator with the default settings:
    var uid = require('rand-token').uid;

    // Generate a 16 character alpha-numeric token:
    var token = uid(16)

To generate *mostly* sequential ids:
    
    // Create a token generator with the default settings:
    var suid = require('rand-token').suid;

    // Generate a 24 (16 + 8) character alpha-numeric token:
    var token = suid(16)

To generate only lower case letters (a-z):

    // Create a token generator with the default settings:    
    var randtoken = require('rand-token').generator({
      chars: 'a-z'
    });

    // Generate a 16 character token:
    var token = randtoken.generate(16);

Alternatively, you can create a generator with the default options and pass the characters to use as the second parameter to `generate`:

    // Create a token generator with the default settings:    
    var randtoken = require('rand-token').generator();

    // Generate a 16 character token:
    var token = randtoken.generate(16, "abcdefghijklnmopqrstuvwxyz");

To generate only upper case letters with `crypto.randomBytes` as the random source:

    var crypto = require('crypto');
    // Create the generator:
    var randtoken = require('rand-token').generator({
      chars: 'A-Z',
      source: crypto.randomBytes 
    });

    // Generate a 16 character token:
    var token = randtoken.generate(16);


# Dependencies

None.

# License

This plugin is released under the MIT license. See the file [LICENSE](LICENSE).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.1] - 2020-03-23
### Added
- None.
### Changed
- Add files list to package.json so that test files are excluded from final tarball.
### Breaking Changes
- None.

## [1.0.0] - 2020-03-23 (unpublished)
### Added
- TypeScript definitions matching v0.x module.
### Changed
- Added engines with min version of Node v10. Module continues to work on node versions going back to 0.8 but dev dependencies and tests require a modern version of Node.
### Breaking Changes
- None for core library vs v0.x but the type definitions and engines changes may cause compile errors.


test/test.js
const assert = require('assert');
const crypto = require('crypto');

describe('Create a default random token generater', function() {
  it('should create successfully', function() {
    var randtoken = require('../index.js');
  });

  it('should create tokens', function() {
    var randtoken = require('../index.js');
    randtoken.generate(16);
  });

  it('should create tokens of the specified length', function() {
    var randtoken = require('../index.js');
    var token = randtoken.generate(16);
    assert(token.length, 16);
  });

  it('should create tokens using the uid function', function() {
    var randtoken = require('../index.js');
    var token = randtoken.uid(16);
    assert(token.length, 16);
  });

  it('should create sequentially sorted tokens', function(done) {
    var randtoken = require('../index.js');
    var token = randtoken.suid(16);
    function genAnotherAndCompare() {
      var secondToken = randtoken.suid(16);
      assert(token < secondToken);      
      done();
    };
    setTimeout(genAnotherAndCompare, 10);
  });
});

var randtoken = require('../index.js');

var randSourceToTest = [
  'default',
  'math',
  'crypto',
  crypto.randomBytes
];
var tokenCharsToTest = [
  {chars: 'a', regex: /^a{16}$/ },
  {chars: 'alpha', regex: /^[a-z]{16}$/ },
  {chars: 'ALPHA', regex: /^[A-Z]{16}$/ },
  {chars: 'numeric', regex: /^[0-9]{16}$/ },
];

function nameOf(test) {
  if (typeof(test) === 'function') {
    return test.name;
  }
  return test;
}

randSourceToTest.forEach(function(randSourceTest) {
  tokenCharsToTest.forEach(function(tokenCharTest) {
    describe('Generate a 16 character token with tokeChars=' + tokenCharTest.chars + 
                ' and a randSource=' + nameOf(randSourceTest), function() {
      var generator;
      it('should create a generator successfully', function() {
        generator = randtoken.generator({source: randSourceTest, chars: tokenCharTest.chars});
      });

      it('should generate tokens that match the regex ' + tokenCharTest.regex, function() {
        for(var i=0;i<1000;i++) {
          var token = generator.generate(16);
          assert(tokenCharTest.regex.test(token));
        }
      });
    });
  });
});


index.d.ts
/**
 * Uses the default generator to generate a token of size characters.
 */
export function uid(size: number, chars?: string): string;

/**
 * Alias for uid(...) function.
 */
export const generate: typeof uid;

/**
 * Uses the default generator to generate mostly sequential ids
 * that can be compared with the usual string less-than/greater-than operators.
 */
export function suid(size: number, epoch?: number, prefixLength?: number): string;

interface GeneratorOptions {
    /**
     * Source of randomness.
     *
     * Acceptable values are a string or a synchronous function with the
     * signature `(size: number): Buffer`.
     * 
     * The following string values are also accepted:
     *
     * * `default` - This is a synonym for using the default of `crypto.pseudoRandomBytes`. You do not need to specify `default` and can simply create the generator without this option specified to get the same effect.
     * * `crypto` - This is a synonym for [`crypto.randomBytes`](http://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback). Note that this may throw an error during use if there is not enough entropy.
     * * `math` - This is a synonym for using the Math.random() function.
     */
    source?: 'default' | 'math' | 'crypto' | ((size: number) => Buffer);

    /**
     * Characters to use when generating tokens.
     *
     * The following string values have special meaning:
     *
     * * `default` - default set of token characters which is numbers, lower case letters, and upper case letters (i.e. `0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`)
     * * `a-z` or `alpha` - lower case characters (i.e. `abcdefghijklmnopqrstuvwxyz`)
     * * `A-Z` or `ALPHA` - upper case characters (i.e. `ABCDEFGHIJKLMNOPQRSTUVWXYZ`)
     * * `0-9` or `numeric` - numbers only (i.e. `0123456789`)
     * * `base32` - use characters from the base32 alphabet, specifically `A-Z` and `2-7`
     *
     * Any other string value will be treated as an array of characters to use.
     *
     * Each character in the list has an equal chance of being used. For
     * example if a character is repeated twice, it will appear twice as often
     * in randomly generated tokens. `chars` may be at most 255 characters long.
     */
    chars?: string;
}

interface RandTokenGenerator {
    generate(size: number): string;
}

/**
 * Create a random token generator using the given options.
 */
export function generator(options?: GeneratorOptions): RandTokenGenerator;


index.test-d.ts
import * as randToken from './index';
import { expectAssignable } from 'tsd';

expectAssignable<string>(randToken.uid(10));
expectAssignable<string>(randToken.uid(10, 'abcd1234'));

expectAssignable<string>(randToken.generate(10));
expectAssignable<string>(randToken.generate(10, 'abcd1234'));

expectAssignable<string>(randToken.suid(10));
expectAssignable<string>(randToken.suid(10, 1421452800000));
expectAssignable<string>(randToken.suid(10, 1421452800000, 5));

expectAssignable<string>(randToken.generator().generate(123));
expectAssignable<string>(randToken.generator({
    chars: 'abc1234',
}).generate(123));
expectAssignable<string>(randToken.generator({
    source: 'default',
}).generate(123));
expectAssignable<string>(randToken.generator({
    source: 'crypto',
}).generate(123));
expectAssignable<string>(randToken.generator({
    source: 'math',
}).generate(123));
expectAssignable<string>(randToken.generator({
    source: (size) => Buffer.alloc(size, 42),
}).generate(123));


index.js
(function () {
  "use strict";

  var crypto = require('crypto');
  var assert = require('assert');
  var cryptoRandomBytes = crypto.pseudoRandomBytes || crypto.randomBytes;

  var numeric = '0123456789';
  var alphaLower = 'abcdefghijklmnopqrstuvwxyz';
  var alphaUpper = alphaLower.toUpperCase();
  // NOTE: This is explicitly in sortable order:
  var alphaNumeric = numeric + alphaUpper + alphaLower;
  
  var defaults = {
    "chars": 'default',
    "source": 'default'
  };

  function validateTokenChars(tokenChars) {
    assert(tokenChars);
    assert(typeof(tokenChars) == 'string');
    assert(tokenChars.length > 0);
    assert(tokenChars.length < 256);
  }

  function bufferAlloc(size) {
    if (Buffer.alloc) {
      return Buffer.alloc(size);
    }
    return new Buffer(size);
  }

  function buildGenerator(options) {
    assert(!options || typeof(options) == 'object');
    options = options || {};
    options.chars = options.chars || defaults.chars;
    options.source = options.source || defaults.source;

    // Allowed characters
    switch( options.chars ) {
      case 'default':
        options.chars = alphaNumeric;
        break;
      case 'a-z':
      case 'alpha':
        options.chars = alphaLower;
        break;
      case 'A-Z':
      case 'ALPHA':
        options.chars = alphaUpper;
        break;
      case '0-9':
      case 'numeric':
        options.chars = numeric;
        break;
      case 'base32':
        options.chars = alphaUpper + "234567";
        break;
      default:
        // use the characters as is
    }
    validateTokenChars(options.chars);

    // Source of randomness:
    switch( options.source ) {
      case 'default':
        options.source = cryptoRandomBytes;
        break;
      case 'crypto':
        options.source = crypto.randomBytes;
        break;
      case 'math':
        options.source = function(size) {
          var buf = bufferAlloc(size);
          for(var i=0;i<size;i++) {
            buf.writeUInt8(Math.floor(256 * Math.random()), i);
          }
          return buf;
        };
        break;
      default:
        assert(typeof(options.source) == 'function');
    }

    return {
      "generate": function(size, chars) {
        if( chars ) {
          validateTokenChars(chars);
        } else {
          chars = options.chars;
        }
        var max = Math.floor(256 / chars.length) * chars.length;
        var ret = "";
        while( ret.length < size ) {
          var buf = options.source(size - ret.length);
          for(var i=0;i<buf.length;i++) {
            var x = buf.readUInt8(i);
            if( x < max ) {
              ret += chars[x % chars.length];
            }
          }
        }
        return ret;
      }
    };
  }

  function base62(n) {
    assert(n >= 0);
    n = Math.floor(n);
    var ret = [];
    do {
      var index = n % 62;
      ret.push(alphaNumeric[index]);
      n = Math.floor(n / 62);
    } while( n > 0);
    return ret.reverse().join("");
  }

  // Default epoch of "2000-01-01T00:00:00+00:00"
  var defaultEpoch = 946684800000;
  var defaultPrefixLength = 8;
  function suidPrefix(epoch, prefixLength) {
    var ret = base62(Date.now() - epoch);
    while( ret.length < prefixLength ) {
      ret = "0" + ret;
    }
    return ret;
  }

  var defaultGenerator = buildGenerator();

  module.exports = {
    generator: buildGenerator,
    generate: defaultGenerator.generate,
    uid: defaultGenerator.generate,
    suid: function(length, epoch, prefixLength) {
      epoch = epoch || defaultEpoch;
      prefixLength = prefixLength || defaultPrefixLength;
      return suidPrefix(epoch, prefixLength) + defaultGenerator.generate(length);
    }
  };
})();
